README.md
# Jenkins i Docker-konteiner

Dette er en prekonfigurert Jenkins-instans som benytter pipeline-as-code-konseptet.

## Tilgjengelige verktøy

Følgende verktøy er tilgjengelig for byggejobber direkte fra `PATH`:
* Java Development Kit 8 (`java`)
* Maven 3.3 (`mvn`)
* Docker klient 1.12 (`docker`)
* Docker Machine 0.8 (`docker-machine`)
* AWS CLI (`aws`)

## Hvordan kjøre applikasjonen

Følgende krav stilles til vertsmaskinen:
* Docker Engine (1.12.1 eller nyere) er installert.
* Det finnes en bruker `jenkins`.

For byggejobber som trenger Docker:
* En Docker Engine, enten den samme som Jenkins kjører i, eller en fjern, som er konfigurert for [toveis TLS-aksess](https://docs.docker.com/engine/security/https/).
* Brukeren `jenkins` har en `.docker`-katalog med nødvendige PEM-filer for å aksessere en Docker Engine via TLS.

For byggejobber som trenger SSH-aksess til tjenere:
* Brukeren `jenkins` har en fil `.ssh/known_hosts` med fingeravtrykkene til de tjenerne Jenkins vil kontakte med SSH.
* Miljøvariabelen `SSH_AUTH_SOCK` inneholder stien til en Unix-socket for en kjørende
  [ssh-agent](https://wiki.archlinux.org/index.php/SSH_keys#SSH_agents) som holder autentiseringsnøkler for de tjenerne
  Jenkins vil kontakte med SSH.
  
For byggejobber som trenger AWS:
* Brukeren `jenkins` har en `.aws`-katalog med nødvendig legitimasjon og innstillinger. Legitimasjonen må peke til
  en AWS-bruker som har nødvendige rettigheter til å utføre det byggejobbene trenger
  (f.eks. en IAM-bruker med sikkerhetspolicien `AmazonEC2FullAccess`).

Applikasjonen kan kjøres som en tjeneste på en Docker-sverm. `pipeline/application.sh create` gjør dette for deg.

Alternativt kan en enkelt konteiner startes på følgende måte:
```
$ USER=jenkins
$ docker run -d \
  -p 80:8080 \
  -v $SSH_AUTH_SOCK:/ssh_auth_sock \
  -v $(eval echo ~$USER)/data:/var/jenkins_home/ \
  -v $(eval echo ~$USER)/.ssh/known_hosts:/var/jenkins_home/.ssh/known_hosts \
  -v $(eval echo ~$USER)/.docker:/var/jenkins_home/.docker \
  -v $(eval echo ~$USER)/.aws:/var/jenkins_home/.aws \
  -v $(eval echo ~$USER)/.m2:/var/jenkins_home/.m2 \
  -e uid=`id -u $USER` \
  -e gid=`id -g $USER` \
  -e DOCKER_HOST=tcp://$HOSTNAME:2376 \
  --restart=unless-stopped \
  --name jenkins \
  docker-registry.dmz.local/eid-jenkins
```
*NB! Containere som ikke startes som service, kan ikke kjøre fullstendige deploy-bygg.*

## Hvordan vedlikeholde bildet

### Oppgradere Jenkins-versjon

Dockerfile inneholder følgende linjer:

```
ARG JENKINS_VERSION=2.23
ARG JENKINS_SHA=6c47f8f6019b9a2be17662033444ce7eec03f4fa
```

På http://repo.jenkins-ci.org/public/org/jenkins-ci/main/jenkins-war/[versjon]/jenkins-war-[versjon].war finner du siste versjon, og SHA1-sum for denne finner du i http://repo.jenkins-ci.org/public/org/jenkins-ci/main/jenkins-war/[versjon]/jenkins-war-[versjon].war.sha1. Erstatt argumentene over med disse for å bygge bilder med denne versjonen.

### Oppgradere tillegg

Det kjøres en rekke `RUN install-plugin.sh <navn> <versjon>` for de tilleggene som er i bruk. Her kan versjoner oppdateres, og nye tillegg kan også eventuelt legges til.

### Bygge bildet

```
$ docker build -t docker-registry.dmz.local/eid-jenkins .
```

### Publisere bildet

```
$ docker push docker-registry.dmz.local/eid-jenkins
```


doc/Changes.txt
feature/PBLEID-14058: Legge til eid-level1-poc
feature/PBLEID-14049: Oppsett av bygg av id-porten
feature/PBLEID-13994: Opprette byggejobb på jenkins02 for eid-common
feature/PBLEID-13993: Lage bygg som ikke trigges ved push
feature/PBLEID-13717: Legge til authlevel jobb og cucumber/slack plugins
feature/PBLEID-13696: Legge til bygg av Resilience
feature/PBLEID-13426: Sette opp e-post med emailext for bruk i prosjektene slik at man får e-post når bygg feiler
feature/PBLEID-13457: Oppdatere med plugins for at Jenkins skal kunne kjøre opp bygg i docker-containere fra Jenkinsfile
feature/PBLEID-13376: Legge til minid-on-the-fly
feature/PBLEID-13404: Docker 1.13-klient


pipeline/application.sh
#!/usr/bin/env bash

fail() {
    ret=$?
    message=${1-"[Failed (${ret})]"}
    echo ${message}
    return ${ret}
}

warn() {
    message=$1
    echo ${message}
}

ok() {
    echo "[OK]"
}

dotSleep() {
    length=${1-1}
    echo -n "."
    sleep ${length};
}

requireArgument() {
    test -z ${!1} && fail "Missing argument '${1}'"
}

image() {
    service=$1
    version=${2-'latest'}
    requireArgument 'service'
    case "${service}" in
        "jenkins")
            image="docker-registry.dmz.local/eid-jenkins:${version}"
            ;;
        *)
            fail "Unknown service ${service}"
    esac
    echo ${image}
}

createService() {
    service=$1
    version=${2-'latest'}
    requireArgument 'service'
    network='pipeline'
    echo -n "Creating service ${service} of version ${version}... "
    image=$(image ${service} ${version})
    case ${service} in
    jenkins)
        output=$(sudo docker service create \
            --network ${network} \
            --constraint 'node.role == manager' \
            --mount type=bind,src=$SSH_AUTH_SOCK,target=/ssh_auth_sock \
            --mount type=bind,src=$(eval echo ~$USER)/data,target=/var/jenkins_home \
            --mount type=bind,src=$(eval echo ~$USER)/.ssh/known_hosts,target=/var/jenkins_home/.ssh/known_hosts \
            --mount type=bind,src=$(eval echo ~$USER)/.docker,target=/var/jenkins_home/.docker \
            --mount type=bind,src=$(eval echo ~$USER)/.m2,target=/var/jenkins_home/.m2 \
            --mount type=bind,src=$(eval echo ~$USER)/.aws,target=/var/jenkins_home/.aws \
            --secret minidonthefly-shenzi \
            -e DOCKER_HOST=tcp://$(hostname -f):2376 \
            -e uid=`id -u $USER` \
            -e gid=`id -g $USER` \
            --name ${service} \
            -p 80:8080 \
            ${image}) \
            || fail "Failed to create service ${service}"
        ;;
    esac
    ok
}

updateService() {
    service=$1
    version=${2-'latest'}
    requireArgument 'service'
    echo -n "Updating service ${service} to version ${version}... "
    image=$(image ${service} ${version})
    output=$(sudo docker service inspect ${service}) || { echo "Service needs to be created"; createService ${service} ${version}; return; }
    output=$(sudo docker service update --image ${image} ${service}) \
        && ok || fail
}

waitForServiceUpdateToComplete() {
    service=$1
    requireArgument 'service'
    echo -n "Waiting for service \"${service}\" to be updated"
    for i in $(seq 1 100); do isServiceUpdateCompleted ${service} && status=true && break || dotSleep; done
    echo -n " "
    ${status} && ok || fail
}

isServiceUpdateCompleted() {
    service="${1}"
    requireArgument 'service'
    [ "$(sudo docker service inspect ${service} -f '{{.UpdateStatus.State}}')" == "completed" ]
}

deleteService() {
    service=$1
    requireArgument 'service'
    echo -n "Deleting service ${service}... "
    output=$(sudo docker service rm ${service}) \
        && ok || fail
}

createNetwork() {
    network=$1
    requireArgument 'network'
    echo -n "Creating network ${network}... "
    output=$(sudo docker network create -d overlay --subnet 10.0.1.0/24 ${network}) \
        && ok || fail
}

deleteNetwork() {
    network=$1
    requireArgument 'network'
    echo -n "Deleting network ${network}... "
    output=$(sudo docker network rm ${network}) \
        && ok || fail
}

waitForServiceToBeAvailable() {
    service=$1
    requireArgument 'service'
    host=${2-'localhost'}
    echo -n "Waiting for service \"${service}\" to be available: "
    status=false
    for i in $(seq 1 200); do
        isServiceAvailable ${service} ${host}
        ret=$?
        [ ${ret} -eq 7 -o ${ret} -eq 27 ] && dotSleep; # Connect failure or request timeout
        [ ${ret} -eq 0 ] && { status=true; break; }
        [ ${ret} -eq 1 ] && break # Unknown service
    done
    ${status} && ok || fail
}

isServiceAvailable() {
    service=$1
    requireArgument 'service'
    host=${2-'localhost'}
    case "${service}" in
        'jenkins')
            url="http://${host}:80"
            ;;
        *)
            echo -n "Unknown service \"${service}\""
            return 1
    esac
    curl -s ${url} --connect-timeout 3 --max-time 10 > /dev/null
}

create() {
    version=${1-'latest'}
    echo "Creating application with version ${version}..."
    createNetwork 'pipeline' || return $?
    createService 'jenkins' ${version} || return $?
    echo "Application created"
}

update() {
    version=${1-'latest'}
    echo "Updating application to version ${version}..."
    updateService 'jenkins' ${version} || return $?
    echo "Application updated"
}

delete() {
    echo "Deleting application..."
    deleteService "jenkins"
    deleteNetwork "pipeline"
    echo "Application deleted"
}

case "${1}" in *)
        function="${1}"
        shift
        ${function} "${@}"
        ;;
esac
